llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -8790,11 +8790,28 @@ SDValue SITargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
                              AMDGPUFunctionArgInfo::LDS_KERNEL_ID);
   }
   case Intrinsic::amdgcn_workitem_id_x:
-    return lowerWorkitemID(DAG, Op, 0, MFI->getArgInfo().WorkItemIDX);
+    if (!MFI->getArgInfo().WorkItemIDX) {
+      // It's undefined behavior if a function marked with the amdgpu-no-*
+      // attributes uses the corresponding intrinsic.
```

Comment:
I thought it was poison? This should return undef, not 0?

Reply:
Those tests don't exhibit undefined cases, so they should not be changed. I think that means the logic here isn't exact, and could be more aggressive if you figure out why the argument isn't present

Reply:
I think I got the order of the handling of the different cases wrong or at least not like GlobalISel does it. I now adapted it to match the behavior of GlobalISel.